### PR TITLE
App Hosting Emulator: Init

### PR DESCRIPTION
**Pull request by mathu97**
Properly initialize the App Hosting emulator when app hosting emulator is selected during `firebase init emulators`. 

When the app hosting emulator is initialized ...
 1. user should be prompted for root directory and `rootDirectory` config should be set
 2. the default start command should be detected and `startCommandOverride` should be set
 3.  user should be prompted to export secrets

Additionally ...
 - Refactored logic used by `apphosting:config:export` command so that it can be re-used at init. 
 - Added scaffolding to make it easy for emulators to add additional prompts when `init emulators` is ran.

### CODE REVIEW

File: src/emulator/initEmulators.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+      default: "./",
+      message: "Specify your app's root directory relative to your repository",
+    });
+
+    return {
+      rootDirectory,
+    };
+    // prompt for apphosting yaml to export
+  },
+  auth: async () => {
```
Comment (joehan) [inline src/emulator/initEmulators.ts:27]: Definitely no need to list out every product here with an empty function - cleaner to just include the products that actually have extra questions, and then change the call site to work with undefined

File: src/apphosting/config.ts [hunk 2]
```diff
@@ -172,6 +176,41 @@ export async function maybeAddSecretToYaml(secretName: string): Promise<void> {
   dynamicDispatch.store(path, projectYaml);
 }
 
+export async function exportConfig(
+  projectId: string,
+  cwd: string,
+  backendRoot: string,
+  environmentConfigFile?: string,
+) {
```
Comment (joehan) [inline src/apphosting/config.ts:184]:
```suggestion
): Promise<void> {
```

File: src/apphosting/config.ts [hunk 3]
```diff
@@ -172,6 +176,41 @@ export async function maybeAddSecretToYaml(secretName: string): Promise<void> {
...
+  for (const [key, value] of secretMaterial) {
+    localAppHostingConfig.addEnvironmentVariable({
+      variable: key,
+      value: value,
+      availability: ["RUNTIME"],
+    });
+  }
+
+  // update apphosting.local.yaml
+  localAppHostingConfig.upsertFile(localAppHostingConfigPath);
```
Comment (joehan) [inline src/apphosting/config.ts:210]:
```suggestion
  await localAppHostingConfig.upsertFile(localAppHostingConfigPath);
```

File: src/apphosting/config.ts [hunk 4]
```diff
@@ -199,3 +238,64 @@ export async function loadConfigForEnvironment(
 
   return envYamlConfig;
 }
+
+/**
+ * Returns the appropriate App Hosting YAML configuration for exporting secrets.
+ *
```
Comment (joehan) [inline src/apphosting/config.ts:244]:
```suggestion
```

File: src/emulator/apphosting/utils.ts [hunk 5]
```diff
@@ -2,6 +2,7 @@ import { pathExists } from "fs-extra";
 import { join } from "path";
```
Comment (joehan) [inline src/emulator/apphosting/utils.ts:2]: High level comment - even though we do it all over the CLI codebase, `utils` files tend to be a antipattern - over time, they fill up with random stuff and that leads to a disorganized codebase. 

Right now, this file has 2 functions that detect what package manager is used. Consider renaming this to 'detectPackageManager.ts' or something to that effect?

File: src/emulator/initEmulators.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    additionalConfigs.set("rootDirectory", backendRelativeDir);
+
+    const backendRoot = join(cwd, backendRelativeDir);
+    try {
+      const startCommand = await detectStartCommand(backendRoot);
+      additionalConfigs.set("startCommandOverride", startCommand);
+    } catch (e) {
+      logger.log(
+        "WARN",
+        "failed to auto-detect your project's start command, consider manually setting the start command by setting the startCommandOverride config",
```
Comment (joehan) [inline src/emulator/initEmulators.ts:36]:
```suggestion
        "Failed to auto-detect your project's start command. Consider manually setting the start command by setting `firebase.json#emulators.apphosting.startCommandOverride`",
```

File: src/apphosting/config.ts [hunk 7]
```diff
@@ -172,6 +176,41 @@ export async function maybeAddSecretToYaml(secretName: string): Promise<void> {
...
+      value: value,
+      availability: ["RUNTIME"],
+    });
+  }
+
+  // update apphosting.local.yaml
+  localAppHostingConfig.upsertFile(localAppHostingConfigPath);
+  logger.info(`Wrote secrets as environment variables to ${APPHOSTING_LOCAL_YAML_FILE}.`);
+}
+
```
Comment (joehan) [inline src/apphosting/config.ts:213]: Nit: Mind cleaning up some of these check warnings while you are here? They seem pretty trivial to tidy up

**Review by joehan — APPROVED** [review]
Some nitty things, but lgtm after some small fixes.

File: src/apphosting/config.ts [hunk 8]
```diff
@@ -172,6 +176,41 @@ export async function maybeAddSecretToYaml(secretName: string): Promise<void> {
...
+    localAppHostingConfig.addEnvironmentVariable({
+      variable: key,
+      value: value,
+      availability: ["RUNTIME"],
+    });
+  }
+
+  // update apphosting.local.yaml
+  localAppHostingConfig.upsertFile(localAppHostingConfigPath);
+  logger.info(`Wrote secrets as environment variables to ${APPHOSTING_LOCAL_YAML_FILE}.`);
```
[the remark this answers is not in the document]
Comment (mathu97) [reply src/apphosting/config.ts:211]: This is where we'd automatically add apphosting.local.yaml files. Doing this in a follow up PR: https://github.com/firebase/firebase-tools/pull/7963

File: src/emulator/apphosting/utils.ts [hunk 9]
```diff
@@ -24,5 +25,20 @@ export async function discoverPackageManager(rootdir: string): Promise<PackageMa
...
+  throw new FirebaseError("Unsupported package manager");
+}
+
+export async function detectStartCommand(rootDir: string) {
+  try {
+    const packageManager = await detectPackageManager(rootDir);
+    return `${packageManager} run dev`;
+  } catch (e) {
+    throw new FirebaseError(
+      "Failed to detect your project's package manager, consider manually setting the start command with the `startCommandOverride` config. ",
```
[the remark this answers is not in the document]
Comment (mathu97) [reply src/emulator/apphosting/utils.ts:41]: done!

File: src/emulator/apphosting/serve.ts [hunk 10]
```diff
@@ -60,7 +61,14 @@ async function serve(port: number, startCommand?: string): Promise<void> {
     return;
   }
 
-  const packageManager = await discoverPackageManager(rootDir);
+  let packageManager: PackageManager = "npm";
```
[the remark this answers is not in the document]
Comment (mathu97) [reply src/emulator/apphosting/serve.ts]: This has been updated. We don't default to `npm` anymore. An error is thrown if lock files for any of the supported package managers are found.

File: src/apphosting/config.ts [hunk 11]
```diff
@@ -199,3 +238,64 @@ export async function loadConfigForEnvironment(
...
+      );
+    }
+
+    userGivenConfigFilePath = allConfigs.get(userGivenConfigFile)!;
+  } else {
+    userGivenConfigFilePath = await promptForAppHostingYaml(
+      allConfigs,
+      "Which environment would you like to export secrets from Secret Manager for?",
+    );
+  }
```
[the remark this answers is not in the document]
Comment (mathu97) [reply src/apphosting/config.ts:277]: ya I moved it to re-use implementation during firebase init. This feels more maintainable but maybe less readable for the reasons you had mentioned. Maybe we can reconsider this when we add support for server side env vars.

**Review by taeold — APPROVED** [review]

File: src/apphosting/config.ts [hunk 12]
```diff
@@ -172,6 +176,41 @@ export async function maybeAddSecretToYaml(secretName: string): Promise<void> {
...
+
+  const configToExport = await loadConfigToExport(cwd, environmentConfigFile);
+  const secretsToExport = configToExport.secrets;
+  if (!secretsToExport) {
+    logger.warn("No secrets found to export in the chosen App Hosting config files");
+    return;
+  }
+
+  const secretMaterial = await fetchSecrets(projectId, secretsToExport);
```
[the remark this answers is not in the document]
Comment (mathu97) [reply src/apphosting/config.ts:234]: addressing this in a follow up pr: https://github.com/firebase/firebase-tools/pull/7973

**Pull request merged**